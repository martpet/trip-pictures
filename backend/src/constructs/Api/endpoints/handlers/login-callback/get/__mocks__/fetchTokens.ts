import { IdTokenPayload, OauthTokens } from '~/constructs/Api/types';

export const dummyOauthTokens: OauthTokens = {
  accessToken: 'dummyAccessToken',
  refreshToken: 'dummyRefreshToken',
  idToken: 'dummyIdToken',
  expiresIn: 123,
  idTokenPayload: {
    sub: 'dummySub',
    nonce: 'dummyIdTokenNonce',
    aud: 'dummyAud',
  } as IdTokenPayload,
};

export const fetchTokens = vi
  .fn()
  .mockName('fetchTokens')
  .mockResolvedValue(dummyOauthTokens);
