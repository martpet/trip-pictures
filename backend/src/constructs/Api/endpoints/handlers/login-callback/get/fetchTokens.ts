import camelcaseKeys from 'camelcase-keys';
import {
  authPaths,
  FetchTokensResponse,
  getIdTokenPayload,
  OauthTokens,
} from 'lambda-layer';
import fetch from 'node-fetch';

type FetchTokensProps = {
  code: string;
  codeVerifier: string;
  clientId: string;
  authDomain: string;
  loginCallbackUrl: string;
};

export const fetchTokens = async ({
  code,
  codeVerifier,
  clientId,
  authDomain,
  loginCallbackUrl,
}: FetchTokensProps): Promise<OauthTokens> => {
  const response = await fetch(`https://${authDomain}${authPaths.token}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      grant_type: 'authorization_code',
      client_id: clientId,
      code,
      code_verifier: codeVerifier,
      redirect_uri: loginCallbackUrl,
    }),
  });

  const data = (await response.json()) as FetchTokensResponse;

  if ('error' in data) {
    throw new Error(data.error);
  }

  return {
    ...camelcaseKeys(data),
    idTokenPayload: await getIdTokenPayload(data.id_token),
  };
};
