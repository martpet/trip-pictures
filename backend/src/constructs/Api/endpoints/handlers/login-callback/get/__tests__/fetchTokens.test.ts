import camelcaseKeys from 'camelcase-keys';
import fetch, { Response } from 'node-fetch';

import {
  getIdTokenPayload,
  itCalls,
  itRejects,
  itResolves,
} from '~/constructs/Api/utils';

import { fetchTokens } from '../fetchTokens';

vi.mock('node-fetch');
vi.mock('camelcase-keys');
vi.mock('~/constructs/Api/utils/getIdTokenPayload');

const args = [
  {
    code: 'dummyCode',
    codeVerifier: 'dummyCodeVerifier',
    clientId: 'dummyClientId',
    authDomain: 'dummyAuthDomain',
    loginCallbackUrl: 'dummyLoginCallbackUrl',
  },
] as unknown as Parameters<typeof fetchTokens>;

vi.mocked(fetch).mockResolvedValue({
  json: () => Promise.resolve({ id_token: 'dummyIdToken' }),
} as Response);

describe('fetchTokens', () => {
  itCalls(fetch, fetchTokens, args);
  itCalls(camelcaseKeys, fetchTokens, args);
  itCalls(getIdTokenPayload, fetchTokens, args);
  itResolves(fetchTokens, args);

  describe('when "fetch" response contains an "error" prop', () => {
    const error = { error: 'dummyErrorMessage' };
    const response = { json: () => Promise.resolve(error) } as Response;
    beforeEach(() => {
      vi.mocked(fetch).mockResolvedValueOnce(response);
    });
    itRejects(fetchTokens, args);
  });
});
