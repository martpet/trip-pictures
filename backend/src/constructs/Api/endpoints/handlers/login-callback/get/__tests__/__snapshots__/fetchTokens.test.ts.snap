// Vitest Snapshot v1

exports[`fetchTokens > calls "camelcase-keys" with correct args 1`] = `
[
  [
    {
      "id_token": "dummyIdToken",
    },
  ],
]
`;

exports[`fetchTokens > calls "fetch" with correct args 1`] = `
[
  [
    "https://dummyAuthDomain/oauth2/token",
    {
      "body": URLSearchParams {
        Symbol(query): [
          "grant_type",
          "authorization_code",
          "client_id",
          "dummyClientId",
          "code",
          "dummyCode",
          "code_verifier",
          "dummyCodeVerifier",
          "redirect_uri",
          "dummyLoginCallbackUrl",
        ],
        Symbol(context): null,
      },
      "headers": {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`fetchTokens > calls "getIdTokenPayload" with correct args 1`] = `
[
  [
    "dummyIdToken",
  ],
]
`;

exports[`fetchTokens > resolves with a correct value 1`] = `
{
  "dummyCamelCaseKey": "dummyCamelCaseValue",
  "idTokenPayload": {
    "aud": "dummyAud",
    "email": "dummyEmail",
    "familyName": "dummyFamilyName",
    "givenName": "dummyGivenName",
    "nonce": "dummyNonce",
    "picture": "dummyPicture",
    "sub": "dummySub",
  },
}
`;

exports[`fetchTokens > when "fetch" response contains an "error" prop > rejects with a correct value 1`] = `[Error: dummyErrorMessage]`;
